src/hooks/routing/useRouterTrade.ts

```diff
@@ -0,0 +1,141 @@
+import { computeRoutes, transformRoutesToTrade } from 'state/routing/utils'
+
+import { isAutoRouterSupportedChain } from './clientSideSmartOrderRouter'
+
+export const INVALID_TRADE = { state: TradeState.INVALID, trade: undefined }
+
+/**
+ * Returns the best trade by invoking the routing api or the smart order router on the client
+ * @param tradeType whether the swap is an exact in/out
+ * @param routerUrl the base URL of the integrator's auto router API
+ * @param amountSpecified the exact amount to swap in/out
+ * @param otherCurrency the desired output/payment currency
+ */
+export function useRouterTrade<TTradeType extends TradeType>(
+  tradeType: TTradeType,
+  routerUrl?: string,
+  amountSpecified?: CurrencyAmount<Currency>,
+  otherCurrency?: Currency
+): {
+  state: TradeState
+  trade: InterfaceTrade<Currency, Currency, TTradeType> | undefined
+} {
+  const { chainId, library } = useActiveWeb3React()
+  const autoRouterSupported = isAutoRouterSupportedChain(chainId)
+  const isWindowVisible = useIsWindowVisible()
+  // Debounce is used to prevent excessive requests to SOR, as it is data intensive.
+  // Fast user actions (ie updating the input) should be debounced, but currency changes should not.
+  const [debouncedAmount, debouncedOtherCurrency] = useDebounce(
+    useMemo(() => [amountSpecified, otherCurrency], [amountSpecified, otherCurrency]),
+    200
+  )
+  const debouncedAmountSpecified = autoRouterSupported && isWindowVisible ? debouncedAmount : undefined
+
+  const [currencyIn, currencyOut]: [Currency | undefined, Currency | undefined] = useMemo(
+    () =>
+      tradeType === TradeType.EXACT_INPUT
+        ? [debouncedAmountSpecified?.currency, debouncedOtherCurrency]
+        : [debouncedOtherCurrency, debouncedAmountSpecified?.currency],
+    [debouncedAmountSpecified, debouncedOtherCurrency, tradeType]
+  )
+
+  const currencyChainId = currencyIn?.chainId as ChainId
+  if (currencyChainId && !isAutoRouterSupportedChain(currencyChainId)) {
+    throw new Error(`Router does not support this token's chain (chainId: ${currencyChainId}).`)
+  }
+
+  const queryArgs = useRouterArguments({
+    tokenIn: currencyIn,
+    tokenOut: currencyOut,
+    amount: debouncedAmountSpecified,
+    tradeType,
+    routerUrl,
+    provider: library as JsonRpcProvider,
+  })
+
+  const { isFetching, isError, data, currentData } = useGetQuoteQuery(queryArgs ?? skipToken, {
+    pollingInterval: ms`15s`,
+    refetchOnFocus: true,
+  })
+
+  const quoteResult: GetQuoteResult | undefined = useIsValidBlock(Number(data?.blockNumber) || 0) ? data : undefined
+
+  const route = useMemo(
+    () => computeRoutes(currencyIn, currencyOut, tradeType, quoteResult),
+    [currencyIn, currencyOut, quoteResult, tradeType]
+  )
+
+  // get USD gas cost of trade in active chains stablecoin amount
+  const gasUseEstimateUSD = useStablecoinAmountFromFiatValue(quoteResult?.gasUseEstimateUSD) ?? null
+
+  const isSyncing = currentData !== data
```

Comment:
do you understand this bit? I feel like `syncing` may be the wrong word?

I think that currentData !== data means the query args changed.